hetmech/matrix.py

```diff
@@ -123,3 +124,55 @@ def copy_array(matrix, copy=True):
         mat_type = numpy.array
     matrix = mat_type(matrix, dtype=numpy.float64, copy=copy)
     return matrix
+
+
+def categorize(metapath):
+    """
+    Returns the classification of a given metapath as one of
+    a set of metapath types which we approach differently.
+
+    Parameters
+    ----------
+    metapath : hetio.hetnet.MetaPath
+
+    Returns
+    -------
+    classification : string
+        One of ['no_repeats', 'disjoint', 'BAAB', 'BABA', 'other']
+    """
+    metanodes = list(metapath.get_nodes())
+    repeated_nodes = {v for i, v in enumerate(metanodes) if
+                      v in metanodes[i + 1:]}
+
+    if not repeated_nodes:
+        return 'no_repeats'
+
+    repeats_only = [node for node in metanodes if node in repeated_nodes]
+
+    # Group neighbors if they are the same
+    grouped = [list(v) for k, v in itertools.groupby(repeats_only)]
+
+    # Handle multiple disjoint repeats, any number, ie. AA,BB,CC,DD,...
+    if len(grouped) == len(repeated_nodes):
+        return 'disjoint'
+
+    # Group [A, BB, A] or [A, B, A, B] into one
+    if len(repeats_only) - len(grouped) <= 1:
+        grouped = [[item for sublist in grouped for item in sublist]]
+
+    # Categorize the reformatted metapath
+    if len(grouped) == 1 and len(grouped[0]) == 4:
+        if grouped[0][0] == grouped[0][-1]:
+            return 'BAAB'
+        else:
+            return 'BABA'
+    else:
+        # Multi-repeats that aren't disjoint, eg. ABCBAC
+        if len(repeated_nodes) > 2:
+            raise NotImplementedError(
+                "Only two overlapping repeats currently supported")
+
+        if len(metanodes) > 5:
```

Comment:
The only reason I do is because we don't have to add any more functionality to correctly categorize `AABBCCDD...`, an arbitrary number of disjoint repeats. Most of the code above that is just to adequately cover the disjoint case. It could be move up somewhat, though, if you think it would be useful.